### PR TITLE
announce upgrades and warn on downgrades

### PR DESCRIPTION
**Pull request by subdavis**
Give users more confidence that they're using the correct instance of DIVE desktop.

## How to test:

Open localstorage in devtools (Application -> Localstorage) and manually edit the value of `desktop.currentVersion` to mock upgrades and downgrades.

### CODE REVIEW

File: client/platform/desktop/frontend/components/Recent.vue [hunk 1]
```diff
@@ -233,6 +237,34 @@ export default defineComponent({
...
+        </v-alert>
+        <v-alert
+          v-if="downgradedVersion"
+          type="warning"
+          color="warning darken-1"
+        >
+          <h3>
+            Downgrade detected
+          </h3>
+          You're using {{ downgradedVersion }}, but  a newer version has been launched before.
```
Comment (BryonLewis) [inline client/platform/desktop/frontend/components/Recent.vue:266]: extra space? `but  a`?

File: client/platform/desktop/frontend/store/settings.ts [hunk 2]
```diff
@@ -66,6 +94,13 @@ async function init() {
   }
   settings.value = settingsValue;
   ipcRenderer.send('update-settings', settings.value);
+
+  /* Populate last acknowledged version */
+  const lastKnownVersion = window.localStorage.getItem(VersionKey);
+  if (lastKnownVersion !== null) {
+    knownVersion.value = lastKnownVersion;
```
Comment (BryonLewis) [inline client/platform/desktop/frontend/store/settings.ts:101]: I may be a little mixed up here following this.  So correct me if I'm wrong.  
1. **`knowVersion` is always retrieved from the `VersionKey` or by default it is null.**  
2. `VersionKey` is only set when you `acknowledgeVersion`.  
3. `acknowledgeVersion` is only triggered in the alert inside of the Vue component by clicking on it
4. `acknowledgeVersion` is inside of a v-alert which require `upgradedVersion` to be truthy
5. `upgradedVersion` is only truthy if `knownVersion` is truthy and it meets the semver requirements

How does the `knownVersion` initially get set or the `VersionKey` localStorage if the value is initially empty?  If it's null shouldn't it be set to the `remote.app.getVersion()`.  I.E.  When a user is coming from a version which has no desktop.currentVersion, how is the value ever set?

**Review by BryonLewis — CHANGES_REQUESTED** [review]
Just some clarification on the initialization of this for a user who has no previous `desktop.currentVersion`

File: client/platform/desktop/frontend/store/settings.ts (hunk 2 above)
Comment (subdavis) [reply client/platform/desktop/frontend/store/settings.ts:101]: 1. yes
2. yes
3. yes
4. yes
5. yes

And yes, it doesn't work, I was changing state as I wrote this, and the end state I got to is unreachable by this code.

> If it's null shouldn't it be set to the remote.app.getVersion()

Yes.

File: client/platform/desktop/frontend/store/settings.ts [hunk 3]
```diff
@@ -1,15 +1,47 @@
...
+
+/**
+ * upgradedVersion indicates that the currently launched instance
+ * is an upgrade since the last time the user acknowledged an instnace of dive.
+ */
+const upgradedVersion = computed(() => {
+  const known = knownVersion.value;
+  if (known === null) {
+    // all versions are greater than null
+    return currentVersion;
```
Comment (BryonLewis) [inline client/platform/desktop/frontend/store/settings.ts:26]: This is minor, If you don't acknowledge this through dismissing and then go to a lower version number you won't get the downgrade warning message?

**Review by BryonLewis — APPROVED** [review]
Minor thing, up to you.

File: client/platform/desktop/frontend/store/settings.ts (hunk 3 above)
Comment (subdavis) [reply client/platform/desktop/frontend/store/settings.ts:26]: Right.  I could do a different implementation with 2 localstorage variables: the currentVersion that gets set immediately on launch and the `acknowleged` variable which gets set true and then back to false every time the version changes.

It was somewhat of an arbitrary choice but this seemed simpler. If you think it's worth doing the other way, I can.

Comment (BryonLewis) [reply client/platform/desktop/frontend/store/settings.ts:26]: I wouldn't.  This I think is a solution for a different and larger problem and that's upgrading and notifying the user about the upgrades and installing them in a consistent location.

**Pull request merged**